pkgs/tools/nix/nixos-option/nixos-option.nix

```diff
@@ -0,0 +1,134 @@
+        in if e.success && lib.isAttrs value && cond value
+           then recurse path' value
+           else f path' value);
+    in recurse [] set;
+
+  # traverse the option tree along `path` from `root`, returning the option or
+  # attrset at the given location
+  optionByPath = path: root:
+    let
+      into = opt: part:
+        if lib.isOption opt && opt.type.descriptionClass == "composite"
+          then opt.type.getSubOptions []
+        else if lib.isOption opt
+          then throw "Trying to access '${part}' inside ${opt.type.name} option while traversing option path '${lib.showOption path}'"
+        else if lib.isAttrs opt && lib.hasAttr part opt
+          then opt.${part}
+        else
+          throw "Found neither an attrset nor supported option type near '${part}' while traversing option path '${lib.showOption path}'";
+    in
+      lib.foldl into root path;
+
+  toPretty = lib.generators.toPretty { multiline = true; };
+  safeToPretty = x:
+    let e = builtins.tryEval (toPretty x);
+    in if e.success then e.value else "[1;31m«error»[m";
+
+  indent = str: lib.concatStringsSep "\n" (map (x: "  " + x) (lib.splitString "\n" str));
+
+  optionAttrNames = attrs: lib.filter (x: x != "_module") (lib.attrNames attrs);
+
+  ## full, non-recursive mode: print an option from `options`
+  renderAttrs = attrs:
+    "This attribute set contains:\n${lib.concatStringsSep "\n" (optionAttrNames attrs)}";
+
+  renderOption = option: value:
+    let
+      entry = cond: heading: value:
+        lib.optional cond "${heading}:\n${indent value}";
+    in lib.concatStringsSep "\n\n" (lib.concatLists [
+      (entry  true                 "Value"       (toPretty value))
+      (entry (option?default)      "Default"     (toPretty option.default))
+      (entry (option?type)         "Type"        (option.type.description))
+      (entry (option?description)  "Description" (lib.removeSuffix "\n" option.description))
+      (entry (option?example)      "Example"     (toPretty option.example))
+      (entry (option?declarations) "Declared by" (lib.concatStringsSep "\n" option.declarations))
+      (entry (option?files)        "Defined by"  (lib.concatStringsSep "\n" option.files))
+    ]);
+
+  renderFull = entry: configEntry:
+    if lib.isOption entry
+      then renderOption entry configEntry
+    else if lib.isAttrs entry
+      then renderAttrs entry
+    else
+      throw "Found neither an attrset nor option at option path '${lib.showOption path'}'";
+
+  ## recursive mode: print paths and values from `config`
+  renderRecursive = config:
+    let
+      renderShort = n: v: "${lib.showOption (path' ++ n)} = ${safeToPretty v};";
+      mapAttrsRecursive' = safeMapAttrsRecursiveCond (x: !lib.isDerivation x);
+    in
+      if lib.isAttrs config
+      then lib.concatStringsSep "\n" (lib.collect lib.isString (mapAttrsRecursive' renderShort config))
+      else renderShort [] config;
+
+in
+  if !lib.hasAttrByPath path' nixos.config
```

Comment:
Stylistic issue: 

instead of `if !P then A else B`
use `if P then B else A`.

Reply:
Hm, that was intentional since I found it easier to read (kind of as an "early return" of sorts, taking care of the error case first) but I'm happy to rephrase it if it's easier to read without the negation